src/Nethermind/Nethermind.Trie/TrieNode.Visitor.cs

```diff
@@ -289,58 +100,155 @@ static void VisitAllSingleThread(TrieNode currentNode, ref TreePath path, ITreeV
+                TNodeContext childContext = nodeContext.Add((byte)i);
+                if (visitor.ShouldVisit(childContext, child.Keccak!))
+                {
+                    Accept(child, childContext, nodeResolver, ref path, isStorage);
+                }
+            }
+            path.TruncateOne();
+        }
+
+        private void VisitAllMultiThread(TrieNode node, TreePath path, in TNodeContext nodeContext, ITrieNodeResolver trieNodeResolver, bool isStorage)
+        {
+            ArrayPoolList<Task>? tasks = null;
+
+            // we need to preallocate children
+            using ArrayPoolList<TrieNode?> output = new ArrayPoolList<TrieNode>(TrieNode.BranchesCount, TrieNode.BranchesCount);
+            int childCount = node.ResolveAllChildBranch(trieNodeResolver, ref path, output.AsSpan());
+            int handledChild = 0;
+
+            path.AppendMut(0);
+            for (int i = 0; i < TrieNode.BranchesCount; i++)
+            {
+                if (output[i] is null) continue;
+                handledChild++;
+                TrieNode child = output[i];
+                path.SetLast(i);
+                child.ResolveKey(trieNodeResolver, ref path, false);
+                TNodeContext childContext = nodeContext.Add((byte)i);
+                if (visitor.ShouldVisit(childContext, child.Keccak!))
+                {
+                    if (
+                        handledChild < childCount // Not the last node
+                        && _threadLimiter.TryTakeSlot(out ConcurrencyController.Slot returner))
+                    {
+                        tasks ??= new ArrayPoolList<Task>(TrieNode.BranchesCount);
+
+                        tasks.Add(SpawnChildVisit(path, child, childContext, trieNodeResolver, returner, isStorage));
+                    }
+                    else
+                    {
+                        Accept(child, childContext, trieNodeResolver, ref path, isStorage);
+                    }
+                }
+            }
+            path.TruncateOne();
+
+            if (tasks is { Count: > 0 })
+            {
+                Task.WaitAll((ReadOnlySpan<Task>)tasks.AsSpan());
+                tasks.Dispose();
+            }
+        }
+
+        private Task SpawnChildVisit(TreePath path, TrieNode child, TNodeContext childContext, ITrieNodeResolver trieNodeResolver, ConcurrencyController.Slot slotReturner, bool isStorage) =>
+            Task.Run(() =>
+            {
+                using ConcurrencyController.Slot _ = slotReturner;
+
+                Accept(child, childContext, trieNodeResolver, ref path, isStorage);
+            });
+
+        private void AddVisited()
+        {
+            int visitedNodes = Interlocked.Increment(ref _visitedNodes);
+
+            // TODO: Fine tune interval? Use TrieNode.GetMemorySize(false) to calculate memory usage?
+            if (visitedNodes % 100_000_000 == 0)
+            {
+                GC.Collect();
```

Comment:
How does this change memory usage of a node?

Reply:
It should use more I guess, although at the rate of tree iteration nowadays, it should get GCed once every few minute or so, so it should be fine.

Reply:
Can you share some graph of memory usage?

Reply:
![Screenshot_2025-02-25_15-50-40](https://github.com/user-attachments/assets/58be137b-9d8b-4cdc-958e-6296212339bd)

1 Mil, 100 Mil and 10 Mil. Twice peak throughput though.